### PR TITLE
Strongly typed hits in ITS + other minor changes

### PR DESCRIPTION
**Pull request by sawenzel**
This PR provides the replacement of TClonsArrays -> std::vector for hits for ITS as important steps for #610.

some other smaller optimization fixes:
a) use const ref for TVector3
b) use cached reference to TVirtualMC::GetMC

### CODE REVIEW

**Comment by sawenzel** [comment]
@iouribelikov , @shahor02 : I am pushing forward with issue #610. Here are the proposed changes for ITS. Please review and let me know.

I needed to do some partial modifications which will disappear in the final version (e.g., duplication of Digitizer::process). 

I understand that you might be working on those files in parallel. Then I could rebase on your changes or -- if you wish -- you could implement the conversion away from TClonesArray yourself ... but I would like to have this soon -:)

**Comment by iouribelikov** [comment]
@sawenzel Thank you, Sandro.   I am having a look at these changes, and will comment back...

File: Detectors/ITSMFT/ITS/simulation/include/ITSSimulation/Detector.h [hunk 1]
```diff
@@ -296,8 +297,9 @@ class Detector : public o2::Base::Detector
     UInt_t *mChipTypeID;           //! Vector of detector type id
     Int_t *mBuildLevel;            //! Vector of Material Budget Studies
 
-    /// Container for data points
-    TClonesArray *mHitCollection;
+    /// Container for hit data
+    // TClonesArray *mHitCollection;
+    std::vector<o2::ITSMFT::Hit>* mHits;
```
Comment (shahor02) [inline Detectors/ITSMFT/ITS/simulation/include/ITSSimulation/Detector.h]: Should not this be std::unique_ptr<std::vector<o2::ITSMFT::Hit>> mHits; ?
Though we could avoid the pointer and simply have a vector data member

File: Detectors/ITSMFT/ITS/simulation/src/Detector.cxx [hunk 2]
```diff
@@ -88,8 +88,8 @@ Detector::Detector()
     mDetectorThickness(nullptr),
     mChipTypeID(nullptr),
     mBuildLevel(nullptr),
-    mHitCollection(new TClonesArray("o2::ITSMFT::Hit")),
-    
+//    mHitCollection(new TClonesArray("o2::ITSMFT::Hit")),
+    mHits(new std::vector<o2::ITSMFT::Hit>),
```
Comment (shahor02) [inline Detectors/ITSMFT/ITS/simulation/src/Detector.cxx]: and here make_unique?

File: Detectors/ITSMFT/ITS/simulation/src/Detector.cxx [hunk 3]
```diff
@@ -362,9 +364,8 @@ Detector::~Detector()
   delete[] mWrapperZSpan;
   delete[] mWrapperLayerId;
 
-  if (mHitCollection) {
-    mHitCollection->Delete();
-    delete mHitCollection;
+  if (mHits) {
```
Comment (shahor02) [inline Detectors/ITSMFT/ITS/simulation/src/Detector.cxx]: ...

File: Detectors/ITSMFT/common/simulation/include/ITSMFTSimulation/Hit.h [hunk 4]
```diff
@@ -110,33 +110,31 @@ class Hit : public o2::BasicXYZEHit<Float_t,Float_t>
     void SetSrcEvID(int srcID, int evID) {
       /// RS: ATTENTION! this is just a trick until we clarify how the hits from different source are
       // provided and identified. At the moment we just create a combined identifier from eventID
-      // and sourceID and store it TEMPORARILY in the cached Point's TObject UniqueID
-      SetUniqueID( ( srcID << Label::nbitsEvID ) | evID);
+      // and sourceID
+      mLabel = ( srcID << Label::nbitsEvID ) | evID;
```
Comment (shahor02) [inline Detectors/ITSMFT/common/simulation/include/ITSMFTSimulation/Hit.h]: OK, this can probably stay like that

**Comment by iouribelikov** [comment]
@sawenzel The output on the control plots looks strange... I would need to have a closer look.
Please, wait a bit with the merging.

**Comment by sawenzel** [comment]
@iouribelikov : Sure. It's good that you have those plots.

File: Detectors/ITSMFT/test/HitAnalysis/src/HitAnalysis.cxx [hunk 5]
```diff
@@ -225,12 +224,11 @@ void HitAnalysis::ProcessChips()
...
 void HitAnalysis::ProcessHits()
 {
-  for (TIter pointiter = TIter(mHitsArray).Begin(); pointiter != TIter::End(); ++pointiter) {
-    Hit *p = static_cast<Hit *>(*pointiter);
-    auto loc = p->GetPos();
-    auto locS = p->GetPosStart();
-    auto glo = mGeometry->getMatrixL2G(p->GetDetectorID())(loc);
-    auto gloS = mGeometry->getMatrixL2G(p->GetDetectorID())(locS);
+  for (auto& hit : *mHits) {
+    auto loc = hit.GetPos();
```
Comment (shahor02) [inline Detectors/ITSMFT/test/HitAnalysis/src/HitAnalysis.cxx]: It looks like i've introduced a bug here, could you please change the lines 228-231 to
auto loc  = mGeometry->getMatrixL2G^(hit.GetDetectorID())(hit.GetPos()); // global->local end position
auto locS = mGeometry->getMatrixL2G^(hit.GetDetectorID())(hit.GetPosStart()); // global->local start position

File: Detectors/ITSMFT/ITS/simulation/include/ITSSimulation/Detector.h (hunk 1 above)
Comment (sawenzel) [reply Detectors/ITSMFT/ITS/simulation/include/ITSSimulation/Detector.h]: I agree with either suggestion (which is the final goal) .. but for the moment neither works with FairRootManager::RegisterAny. So I am afraid we have to postpone this improvement.

File: Detectors/ITSMFT/ITS/simulation/src/Detector.cxx (hunk 2 above)
Comment (sawenzel) [reply Detectors/ITSMFT/ITS/simulation/src/Detector.cxx]: see above

File: Detectors/ITSMFT/test/HitAnalysis/src/HitAnalysis.cxx (hunk 5 above)
Comment (sawenzel) [reply Detectors/ITSMFT/test/HitAnalysis/src/HitAnalysis.cxx]: will do

**Comment by sawenzel** [comment]
Not sure why the build fails partially....

File: Detectors/ITSMFT/test/HitAnalysis/src/HitAnalysis.cxx (hunk 5 above)
Comment (shahor02) [reply Detectors/ITSMFT/test/HitAnalysis/src/HitAnalysis.cxx]: sorry, my fault, misplaced (), should be
auto loc = mGeometry->getMatrixL2G(hit.GetDetectorID())^(hit.GetPos()); // global->local end position
auto locS = mGeometry->getMatrixL2G(hit.GetDetectorID())^(hit.GetPosStart()); // global->local start position

**Comment by iouribelikov** [comment]
@sawenzel Do not yet know why exactly, but the strange test results originate, in fact, from the commit 8ffbef2 ...

**Comment by sawenzel** [comment]
@iouribelikov : No idea ... but I could take this commit out for the moment.

**Comment by iouribelikov** [comment]
This is exactly what I am trying at the moment: taking out that commit locally...  I'll let you know...

**Comment by iouribelikov** [comment]
@sawenzel Confirmed !   If the commit 8ffbef2 is taken out, all the rest of this PR seems to work fine !
(That commit seems to break something in the propagation of the MC labels hit->digit->cluster->track)

**Comment by iouribelikov** [comment]
@sawenzel Aha ! Here is the bug (and a fix to it):
--- a/Detectors/ITSMFT/common/simulation/include/ITSMFTSimulation/Hit.h
+++ b/Detectors/ITSMFT/common/simulation/include/ITSMFTSimulation/Hit.h
@@ -120,7 +120,7 @@ class Hit : public o2::BasicXYZEHit<Float_t,Float_t>
       // and SrcEv id stored as mLabel
       int srcID = ( mLabel>>Label::nbitsEvID ) & Label::maskSrcID;
       int evID = mLabel & Label::maskEvID;
-      return Label( mLabel, evID, srcID );
+      return Label( GetTrackID(), evID, srcID );
     }

**Comment by sawenzel** [comment]
@iouribelikov : Excellent!! I owe you one.

**Comment by iouribelikov** [comment]
@sawenzel Good. I think we can merge this PR (if it's OK also with Ruben).

**Comment by sawenzel** [comment]
@iouribelikov, @shahor02 : I fixed the problem and squashed some commits.  Thanks for the comments.

**Pull request merged**